cornac/models/gru4rec/recom_gru4rec.py

```diff
@@ -0,0 +1,279 @@
+        )
+
+    def _top1(self, O, Y, M):
+        import torch
+
+        target_scores = torch.diag(O)
+        target_scores = target_scores.reshape(target_scores.shape[0], -1)
+        return torch.sum(
+            (
+                torch.mean(
+                    torch.sigmoid(O - target_scores) + torch.sigmoid(O**2), axis=1
+                )
+                - torch.sigmoid(target_scores**2) / (M + self.n_sample)
+            )
+        )
+
+    def fit(self, train_set, val_set=None):
+        super().fit(train_set, val_set)
+        import torch
+
+        from .gru4rec import GRU4RecModel, IndexedAdagradM, io_iter
+
+        if self.logq and self.loss == "cross-entropy":
+            pop = Counter(self.train_set.uir_tuple[1])
+            self.P0 = torch.tensor(
+                [pop[iid] for (_, iid) in self.train_set.iid_map.items()],
+                dtype=torch.float32,
+                device=self.device,
+            )
+
+        self.model = GRU4RecModel(
+            self.total_items,
+            self.layers,
+            self.dropout_p_embed,
+            self.dropout_p_hidden,
+            self.embedding,
+            self.constrained_embedding,
+        ).to(self.device)
+
+        self.model._reset_weights_to_compatibility_mode()
+
+        opt = IndexedAdagradM(
+            self.model.parameters(), self.learning_rate, self.momentum
+        )
+
+        progress_bar = trange(1, self.n_epochs + 1, disable=not self.verbose)
+        for _ in progress_bar:
+            H = []
+            for i in range(len(self.layers)):
+                H.append(
+                    torch.zeros(
+                        (self.batch_size, self.layers[i]),
+                        dtype=torch.float32,
+                        requires_grad=False,
+                        device=self.device,
+                    )
+                )
+            total_loss = 0
+            cnt = 0
+            for inc, (in_iids, out_iids, start_mask, valid_id) in enumerate(
+                io_iter(
+                    s_iter=self.train_set.s_iter,
+                    uir_tuple=self.train_set.uir_tuple,
+                    n_sample=self.n_sample,
+                    sample_alpha=self.sample_alpha,
+                    rng=self.rng,
+                    batch_size=self.batch_size,
+                    shuffle=True,
+                )
+            ):
+                for i in range(len(H)):
+                    H[i][np.nonzero(start_mask)[0], :] = 0
+                    H[i].detach_()
+                    H[i] = H[i][valid_id]
+                in_iids = torch.tensor(in_iids, requires_grad=False, device=self.device)
+                out_iids = torch.tensor(
+                    out_iids, requires_grad=False, device=self.device
+                )
+                self.model.zero_grad()
+                R = self.model.forward(in_iids, H, out_iids, training=True)
+                L = self.loss_function(R, out_iids, len(in_iids)) / self.batch_size
```

Comment:
I would suggest to move all functions related to defining the loss criteria to the `gru4rec.py`. That would improve code structure and readability quite a bit.

Reply:
Updated.